src/webgpu/shader/validation/functions/alias_analysis.spec.ts

```diff
@@ -0,0 +1,271 @@
+      return `_ = abs(${ref});`;
+    }
+    case 'index_access': {
+      return `{ var arr : array<i32, 4>; _ = arr[${ref}]; }`;
+    }
+    case 'let_init': {
+      return `{ let tmp = ${ref}; }`;
+    }
+    case 'var_init': {
+      return `{ var tmp = ${ref}; }`;
+    }
+    case 'return': {
+      return `{ return ${ref}; }`;
+    }
+    case 'switch_cond': {
+      return `switch(${ref}) { default { break; } }`;
+    }
+    default: {
+      unreachable(`Unhandled usage type`);
+    }
+  }
+}
+
+function shouldPass(aliased: boolean, a_use: Use, b_use: Use): boolean {
+  // Expect fail if the pointers are aliased and at least one of the accesses is a write.
+  // If either of the accesses is a "no access" then expect pass.
+  let should_pass = aliased ? !(a_use.is_write || b_use.is_write) : true;
+  should_pass = a_use.use === 'no_access' ? true : should_pass;
+  should_pass = b_use.use === 'no_access' ? true : should_pass;
+  return should_pass;
+}
+
+g.test('two_pointers')
+  .desc(`Test aliasing of two pointers passed to a function.`)
+  .params(u =>
+    u
+      .combine('address_space', ['private', 'function'] as const)
+      .combine('a_use', kUses)
+      .combine('b_use', kUses)
+      .combine('aliased', [true, false])
+      .beginSubcases()
+  )
+  .fn(t => {
+    const code = `
+${t.params.address_space === 'private' ? `var<private> va : i32; var<private> vb : i32;` : ``}
+
+fn callee(pa : ptr<${t.params.address_space}, i32>,
+          pb : ptr<${t.params.address_space}, i32>) -> i32 {
+  ${generateUse(t.params.a_use.use, `*pa`)}
+  ${generateUse(t.params.b_use.use, `*pb`)}
+  return 0;
+}
+
+fn caller() {
+  ${t.params.address_space === 'function' ? `var va : i32; var vb : i32;` : ``}
+  callee(&va, ${t.params.aliased ? `&va` : `&vb`});
+}
+`;
+    t.expectCompileResult(shouldPass(t.params.aliased, t.params.a_use, t.params.b_use), code);
+  });
+
+g.test('one_pointer_one_module_scope')
+  .desc(`Test aliasing of a pointer with a direct access to a module-scope variable.`)
+  .params(u =>
+    u
+      .combine('a_use', kUses)
+      .combine('b_use', kUses)
+      .combine('aliased', [true, false])
+      .beginSubcases()
+  )
+  .fn(t => {
+    const code = `
+var<private> va : i32;
+var<private> vb : i32;
+
+fn callee(pb : ptr<private, i32>) -> i32 {
+  ${generateUse(t.params.a_use.use, `va`)}
+  ${generateUse(t.params.b_use.use, `*pb`)}
+  return 0;
+}
+
+fn caller() {
+  callee(${t.params.aliased ? `&va` : `&vb`});
+}
+`;
+    t.expectCompileResult(shouldPass(t.params.aliased, t.params.a_use, t.params.b_use), code);
+  });
+
+g.test('subcalls')
+  .desc(`Test aliasing of two pointers passed to a function, and then passed to other functions.`)
+  .params(u =>
+    u
+      .combine(
+        'a_use',
+        kUses.filter(el => ['no_access', 'assign', 'binary_lhs'].includes(el.use))
+      )
+      .combine(
+        'b_use',
+        kUses.filter(el => ['no_access', 'assign', 'binary_lhs'].includes(el.use))
+      )
```

Comment:
With `const kUses: Record<string, Use>` this would be:

```suggestion
      .combine('a_use', ['no_access', 'assign', 'binary_lhs'] as UseName[])
      .combine('b_use', ['no_access', 'assign', 'binary_lhs'] as UseName[])
```